File: src/services/classifierService.js
/**
 * Rule-based & Statutory Legal Classifier for Ayurvedic Innovations
 * Evaluates Indian Patents Act (3p, 3e, 3d) & Biological Diversity Act 2002 (NBA)
 */
export function classifyAyurvedicInvention(profile = {}) {
  const {
    q1_components = '',
    q2_process = '',
    q3_evidence = '',
    q4_source = '',
    q5_intent = '',
  } = profile;

  const text = `${q1_components} ${q2_process} ${q3_evidence} ${q4_source} ${q5_intent}`.toLowerCase();

  const analysis = {
    section3pRisk: 'LOW',
    section3eRisk: 'LOW',
    section3dRisk: 'LOW',
    nbaApprovalRequired: false,
    commercialStrategy: [],
    recommendedActions: [],
    statutoryCitations: [],
  };

  // Section 3(p) Traditional Knowledge Assessment
  const traditionalKeywords = [
    'charaka', 'sushruta', 'ashtanga', 'bhavaprakasha', 'classical', 'traditional',
    'ayurvedic text', 'ancient', 'shastra', 'rasashastra', 'taila', 'arishta', 'asava', 'kwatha', 'churna'
  ];
  const isTraditionalSource = q4_source === 'classical' || traditionalKeywords.some((k) => text.includes(k));

  if (isTraditionalSource) {
    if (!text.includes('novel') && !text.includes('modified') && !text.includes('extract') && !text.includes('nanoparticle')) {
      analysis.section3pRisk = 'HIGH';
      analysis.recommendedActions.push(
        'High Section 3(p) rejection risk: The formulation closely resembles traditional prior art in TKDL. Consider patenting a novel extraction method or proprietary delivery system rather than the crude herbal mixture.'
      );
    } else {
      analysis.section3pRisk = 'MEDIUM';
      analysis.recommendedActions.push(
        'Moderate Section 3(p) risk: Novel modifications exist over traditional texts. Ensure comparative characterization against classical preparation methods is detailed in the specification.'
      );
    }
    analysis.statutoryCitations.push('Patents Act, 1970 — Section 3(p)');
  }

  // Section 3(e) Admixture / Synergy Assessment
  const isPolyherbal = q1_components.includes(',') || text.includes('combination') || text.includes('polyherbal') || text.includes('blend');
  const hasSynergyProof = text.includes('synerg') || text.includes('combination index') || text.includes('isobologram') || text.includes('statistically significant');

  if (isPolyherbal) {
    if (!hasSynergyProof) {
      analysis.section3eRisk = 'HIGH';
      analysis.recommendedActions.push(
        'Section 3(e) Admixture objection likely: You must provide experimental comparative synergy assays showing Combination (A+B) exceeds the sum of individual bioactivities (A alone + B alone).'
      );
    } else {
      analysis.section3eRisk = 'LOW';
      analysis.recommendedActions.push(
        'Synergy data indicated: Quantify synergistic ratio with Combination Index (CI < 1) in your complete patent specification to overcome Section 3(e).'
      );
    }
    analysis.statutoryCitations.push('Patents Act, 1970 — Section 3(e)');
  }

  // Biological Diversity Act (NBA) Assessment
  const usesIndianBioResource = q4_source === 'cultivated' || q4_source === 'wild' || q4_source === 'classical' || text.includes('plant') || text.includes('herb') || text.includes('species') || text.includes('india');

  if (usesIndianBioResource) {
    analysis.nbaApprovalRequired = true;
    analysis.recommendedActions.push(
      'MANDATORY NBA CLEARANCE: Sourcing biological resources from India triggers Section 6 of the Biological Diversity Act, 2002. File Form III with the National Biodiversity Authority (NBA) prior to patent grant.'
    );
    analysis.statutoryCitations.push('Biological Diversity Act, 2002 — Section 3 & Section 6; Form III');
  }

  // Commercial Strategy Formulation
  if (analysis.section3pRisk === 'HIGH' && !hasSynergyProof) {
    analysis.commercialStrategy.push('Trademark Registration (Class 5 & Class 3) for brand name protection');
    analysis.commercialStrategy.push('Trade Secret protection for proprietary processing parameters and temperature controls');
    if (text.includes('region') || text.includes('geographic') || text.includes('origin')) {
      analysis.commercialStrategy.push('Geographical Indication (GI) registration for regional heritage reputation');
    }
  } else {
    analysis.commercialStrategy.push('File Indian Provisional Patent Application (Form 1 & Form 2) to secure Priority Date');
    analysis.commercialStrategy.push('File Form III with National Biodiversity Authority (NBA)');
    analysis.commercialStrategy.push('Complete in vivo synergy & pharmacokinetic studies within 12 months before filing Complete Specification');
    analysis.commercialStrategy.push('PCT International Filing within 12 months for global market expansion');
  }

  return {
    profileSubmitted: profile,
    analysis,
    timestamp: new Date().toISOString(),
  };
}

File: backend/services/classifier_service.py
"""
Statutory rule evaluation for Ayurvedic IP
"""

def evaluate_invention_profile(profile: dict) -> dict:
    flags = []
    recommendations = []
    
    bio_tk = str(profile.get("bioResources", "")).lower()
    disclosure = str(profile.get("disclosure", "")).lower()
    description = str(profile.get("description", "")).lower()
    problem = str(profile.get("problem", "")).lower()
    novelty = str(profile.get("novelty", "")).lower()

    # Rule 1: Section 3(p) & Traditional Knowledge
    if "traditional" in bio_tk or "vedic" in bio_tk or "ayurved" in description or "ayurved" in novelty:
        flags.append({
            "code": "SECTION_3P_TK",
            "title": "Section 3(p) - Traditional Knowledge Bar",
            "severity": "HIGH",
            "message": "Inventions based on traditional knowledge face strict non-patentability objections under Section 3(p). You must establish novel non-obvious technical modifications or distinct synergistic mechanisms beyond classical texts."
        })
        recommendations.append("Conduct a TKDL prior art search before filing to verify classical formulation references.")

    # Rule 2: NBA Form III Clearance (Biological Diversity Act 2002)
    if "biological" in bio_tk or "plant" in bio_tk or "herb" in bio_tk or "extract" in bio_tk:
        flags.append({
            "code": "NBA_SECTION_6",
            "title": "Biological Diversity Act (Section 6 Form III)",
            "severity": "HIGH",
            "message": "Inventions utilizing Indian biological resources mandate prior approval from the National Biodiversity Authority (NBA) under Form III before patent grant."
        })
        recommendations.append("File NBA Form III with the National Biodiversity Authority (NBA Chennai) to prevent Section 64 revocation.")

    # Rule 3: Prior Public Disclosure
    if "yes_public" in disclosure or "public" in disclosure:
        flags.append({
            "code": "NOVELTY_PUBLIC_DISCLOSURE",
            "title": "Novelty Destruction via Prior Disclosure",
            "severity": "CRITICAL",
            "message": "Public disclosure prior to filing priority application in India destroys universal novelty."
        })
        recommendations.append("File priority application immediately or rely on trade secret / trademark protections.")

    # Rule 4: Section 3(e) Synergism
    if "combination" in description or "ratio" in novelty or "admixture" in description:
        flags.append({
            "code": "SECTION_3E_ADMIXTURE",
            "title": "Section 3(e) - Synergism Requirement",
            "severity": "MEDIUM",
            "message": "Polyherbal combinations must demonstrate statistically validated synergistic efficacy (Combination Index < 0.8) to overcome Section 3(e) mere admixture rejection."
        })
        recommendations.append("Prepare comparative pharmacological synergy data showing enhanced bio-efficacy over sum of individual components.")

    return {
        "status": "evaluated",
        "flags": flags,
        "recommendations": recommendations,
        "totalHurdles": len(flags),
    }
